integrations/mcp/src/haystack_integrations/tools/mcp/mcp_tool.py

```diff
@@ -0,0 +1,625 @@
+    """Error when a tool is not found on the server."""
+
+    pass
+
+
+class MCPResponseTypeError(MCPError):
+    """Error when response content type is not supported."""
+
+    pass
+
+
+class MCPInvocationError(ToolInvocationError):
+    """Error during tool invocation."""
+
+    pass
+
+
+class MCPClient(ABC):
+    """
+    Abstract base class for MCP clients.
+
+    This class defines the common interface and shared functionality for all MCP clients,
+    regardless of the transport mechanism used.
+    """
+
+    def __init__(self) -> None:
+        self.session: ClientSession | None = None
+        self.exit_stack: AsyncExitStack = AsyncExitStack()
+        self.stdio: MemoryObjectReceiveStream[types.JSONRPCMessage | Exception] | None = None
+        self.write: MemoryObjectSendStream[types.JSONRPCMessage] | None = None
+
+    @abstractmethod
+    async def connect(self) -> list[Tool]:
+        """
+        Connect to an MCP server.
+
+        :returns: List of available tools on the server
+        :raises MCPConnectionError: If connection to the server fails
+        """
+        pass
+
+    async def call_tool(self, tool_name: str, tool_args: dict[str, Any]) -> Any:
+        """
+        Call a tool on the connected MCP server.
+
+        :param tool_name: Name of the tool to call
+        :param tool_args: Arguments to pass to the tool
+        :returns: Result of the tool invocation
+        :raises MCPError: If not connected to an MCP server
+        :raises MCPInvocationError: If the tool invocation fails
+        :raises MCPResponseTypeError: If response type is not TextContent
+        """
+        if not self.session:
+            message = "Not connected to an MCP server"
+            raise MCPError(message)
+
+        try:
+            result = await self.session.call_tool(tool_name, tool_args)
+
+            # Process and validate the response content
+            processed_result = self._process_response_content(result, tool_name)
+            return processed_result
+
+        except Exception as e:
+            if isinstance(e, MCPError):
+                # Re-raise specific MCP errors
+                raise
+            message = f"Failed to invoke tool '{tool_name}': {e!s}"
+            raise MCPInvocationError(message) from e
+
+    def _process_response_content(self, result, tool_name: str) -> Any:
+        """
+        Process response content from an MCP tool call, accepting only TextContent.
+
+        :param result: CallToolResult from MCP tool call
+        :param tool_name: Name of the called tool (for error messages)
+        :returns: The original CallToolResult object
+        :raises MCPResponseTypeError: If content type is not TextContent
+        :raises MCPInvocationError: If the tool call resulted in an error
+        """
+        # Return the result directly if it's not a CallToolResult structure
+        if not hasattr(result, "content"):
+            return result
```

Comment:
Why do we return a result if it doesn't abide by the `CallToolResult` structure?

Reply:
Mistake 🙏